### PR TITLE
Implement etl

### PR DESCRIPTION
**Pull request by sturmer**

### CODE REVIEW

**Comment by sturmer** [comment]
Hi everyone, I've just submitted an implementation for [ETL](https://github.com/exercism/x-common/blob/master/etl.md). As usual, any comment is welcome.

**Comment by sturmer** [comment]
Dear @yurrriq, @arguello, @bennn, when/if you have time I could use your input on this.

File: etl/etl-test.rkt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  (define suite
+    (test-suite
+     "etl tests"
+     
+     (test-true "empty hash"             
+                (zero? (hash-count (etl (make-hash)))))
+     
+     (test-case
+      "mixed case input"
+      (let ([actual (etl mixed-case-input)])
```
Comment (bennn) [inline etl/etl-test.rkt]: Could also assert `(check-equal? (hash-count actual) (hash-count expected))`

File: etl/example.rkt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#lang racket
+
+(provide (contract-out
+          [etl (-> hash? hash?)]))          
+
+; TODO Provide contract on the keys and values of the input and output
+
+(define (etl h)
```
Comment (bennn) [inline etl/example.rkt]: Looks good, but you can use `for*/hash` to build the result and `in-hash` to walk the old table faster.

```
(define (etl h)
    (for*/hash ([(score letter*) (in-hash h)]
                [letter (in-list letter*)])
```

File: etl/example.rkt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#lang racket
+
+(provide (contract-out
+          [etl (-> hash? hash?)]))          
+
+; TODO Provide contract on the keys and values of the input and output
```
Comment (bennn) [inline etl/example.rkt]: Probably `(hash/c exact-nonnegative-integer? (listof string?)) ...`, but it'd be cool to make contracts for lowercase & uppercase letters.

File: etl/etl-test.rkt (hunk 1 above)
Comment (sturmer) [inline etl/etl-test.rkt]: Good idea, will do.

File: etl/example.rkt (hunk 3 above)
Comment (sturmer) [inline etl/example.rkt]: I'll try this one, but I can't guarantee enough Racketeerism to get acceptable results :)

File: etl/example.rkt (hunk 2 above)
Comment (sturmer) [inline etl/example.rkt]: Those I didn't know: duly noted, thanks.

File: etl/example.rkt (hunk 3 above)
Comment (bennn) [inline etl/example.rkt]: :)

**Comment by sturmer** [comment]
@bennn Made some honest contracts working. What do you think?

**Comment by bennn** [comment]
They're good!

File: etl/etl-test.rkt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      (define actual (etl mixed-case-input))
+      (check-equal? (hash-count actual) (hash-count expected))
+      (for ([k (hash-keys actual)])
+        #;(printf "actual: ~a -> ~a, expected: ~a -> ~a\n"
+                  k (hash-ref actual k)
+                  k (hash-ref expected k))
+        (check-equal? (hash-ref actual k)
+                      (hash-ref expected k))))
+
+     (test-true
```
Comment (bennn) [inline etl/etl-test.rkt]: A shorter way is `(test-exn "test bad input: negative keys" exn:fail:contract? (lambda () (etl negative-keys input)))`

Comment (sturmer) [inline etl/etl-test.rkt]: Thanks for the comments @bennn. That would also get rid of the ``(lambda _ #t)`` which I didn't love --- I'll admit it. I am going to make those changes tomorrow.

**Comment by sturmer** [comment]
The contract testing part is now more concise, thanks again to @bennn.

**Comment by sturmer** [comment]
If this is considered good, could anyone please merge it? I think @arguello was so kind to do it last time? Please let me know if there is anything else I can do on my side.

File: etl/example.rkt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  (hash/c exact-nonnegative-integer? (listof string?)))
+
+;; Keys will always be lower-case
+(define valid-output
+  (hash/c lower-case? exact-nonnegative-integer?))
+
+(define (etl h)
+  (for*/hash ([(score letter*) (in-hash h)]
+              [letter (in-list letter*)])
+    ;(printf "~a: ~a\n" letter score)
```
Comment (yurrriq) [inline etl/example.rkt]: I might suggest deleting this line.

**Comment by yurrriq** [comment]
After a quick once over it LGTM. Dig the contracts.

File: etl/example.rkt (hunk 5 above)
Comment (sturmer) [inline etl/example.rkt]: Approved!

**Comment by sturmer** [comment]
@yurrriq Commented line removed. Thanks for digging the contracts :)

**Comment by sturmer** [comment]
@yurrriq Lest this might escape your radar, a small reminder.

**Comment by sturmer** [comment]
Thank you @arguello!

**Comment by arguello** [comment]
Thank you :)

**Comment by yurrriq** [comment]
Thanks, guys.

**Pull request merged**